File: frontend/src/lib/rbac.ts
// Frontend mirror of backend RBAC (Section 9) — used to hide actions/nav the user
// can't perform. The backend remains the source of truth and re-checks every mutation.
const FULL = new Set(["admin", "master_user"]);
const SUPERVISOR = new Set([...FULL, "technical_supervisor"]);
const TECH = new Set([...SUPERVISOR, "technician"]);
const OUTREACH = new Set([...FULL, "motivation"]);
const RECEPTION = new Set([...TECH, "general"]);

// Mirror of backend WRITE_ROLES (app/core/rbac.py). The backend stays the source of
// truth and re-checks every mutation; this map only decides what the UI hides/disables.
const WRITE_ROLES: Record<string, Set<string>> = {
  // identity / config — master only
  orgs: FULL,
  staff: FULL,
  settings: FULL,
  // data entry — technician and up
  bags: TECH,
  components: TECH,
  qc: TECH,
  store: TECH,
  vehicles: TECH,
  hospitals: TECH,
  patients: TECH,
  thalassemia: TECH,
  therapeutic: TECH,
  inquiries: TECH,
  organisations: TECH,
  // outreach
  donors: new Set([...TECH, "motivation"]),
  donations: new Set([...TECH, "motivation"]),
  camps: new Set([...TECH, "motivation"]),
  feedback: OUTREACH,
  // reception
  "blood-requests": RECEPTION,
  invoices: RECEPTION,
  reception: RECEPTION,
  // tools
  barcodes: TECH,
  labels: TECH,
  reservations: TECH,
  downloads: TECH,
  "custom-reports": SUPERVISOR,
  "custom-registers": SUPERVISOR,
  // lab workflow stage transitions
  discard: SUPERVISOR,
  shift: SUPERVISOR,
  validation: SUPERVISOR,
};
const DEFAULT_WRITE = TECH;

export function canWrite(role: string | undefined, resource: string): boolean {
  if (!role) return false;
  // normalise a REST path ("/hospitals", "donors/") to its resource tag
  const tag = resource.replace(/^\/+|\/+$/g, "").split("/")[0];
  return (WRITE_ROLES[tag] ?? DEFAULT_WRITE).has(role);
}

export function roleLabel(role?: string): string {
  return (
    {
      master_user: "Master User",
      admin: "Administrator",
      technician: "Technician",
      technical_supervisor: "Technical Supervisor",
      motivation: "Motivation",
      general: "General",
    }[role ?? ""] ?? role ?? "—"
  );
}

File: backend/app/core/rbac.py
"""Role-based access control (Section 9).

Roles: admin / master_user (full), technical_supervisor (validate + discard + all
technician), technician (data entry on pipelines / QC / bags / donors / camps),
motivation (donor & camp outreach, recall, feedback), general (read everything +
reception). Reads are open to any authenticated user; writes are gated per resource.
"""
from fastapi import Depends, HTTPException

from app.core.deps import get_current_user
from app.models.identity import User

FULL = {"admin", "master_user"}
SUPERVISOR = FULL | {"technical_supervisor"}
TECH = SUPERVISOR | {"technician"}
OUTREACH = FULL | {"motivation"}
RECEPTION = TECH | {"general"}

# resource tag -> set of designations allowed to mutate it
WRITE_ROLES: dict[str, set[str]] = {
    # identity / config — master only
    "orgs": FULL,
    "staff": FULL,
    "settings": FULL,
    # data entry — technician and up
    "bags": TECH,
    "components": TECH,
    "qc": TECH,
    "store": TECH,
    "vehicles": TECH,
    "hospitals": TECH,
    "patients": TECH,
    "thalassemia": TECH,
    "therapeutic": TECH,
    "inquiries": TECH,
    # outreach
    "donors": TECH | {"motivation"},
    "donations": TECH | {"motivation"},
    "camps": TECH | {"motivation"},
    "feedback": OUTREACH,
    # reception
    "blood-requests": RECEPTION,
    "invoices": RECEPTION,
    # tools
    "barcodes": TECH,
    "labels": TECH,
    "reservations": TECH,
    "downloads": TECH,
    "custom-reports": SUPERVISOR,
    "custom-registers": SUPERVISOR,
}
DEFAULT_WRITE = TECH


def role_of(user: User) -> str:
    return "master_user" if user.is_master_user else user.designation


def require_write(resource: str):
    allowed = WRITE_ROLES.get(resource, DEFAULT_WRITE)

    def _dep(user: User = Depends(get_current_user)) -> User:
        if role_of(user) not in allowed:
            raise HTTPException(403, f"Your role ({role_of(user)}) cannot modify {resource}")
        return user

    return _dep


def require_roles(*roles: str):
    allowed = set(roles) | FULL

    def _dep(user: User = Depends(get_current_user)) -> User:
        if role_of(user) not in allowed:
            raise HTTPException(403, "Insufficient role for this action")
        return user

    return _dep
